### ISSUE TITLE
Document how to extract message body using firebase_messaging

### ISSUE THREAD

**Issue by kbrmimbyl** [flutter/flutter#28196]
There's some confusion w.r.t. whether `onLaunch, onResume` and `onMessage` are getting called on iOS or not. Many people are still complaining that one or more of these are not getting called (for e.g. see #18425).

## How to reproduce?

Add `firebase_messaging` to your flutter project, configure for both iOS and Android as explained in the plugin's README. Then send a notification each to your iOS and Android apps using Firebase Console> Cloud Messaging by setting only the "Notification Text" in Step 1 (this is the only mandatory field in this step). Remember to set `click_action` to `FLUTTER_NOTIFICATION_CLICK` under Additional Options (optional, Step 5). In your flutter project, try to extract what you put in as "Notification Text" above using the following dart code to configure `FirebaseMessaging`:

```dart
onMessage: (Map<String, dynamic> msg) {
  _message = msg['notification']['body'];
  print(_message);
},
onResume: (Map<String, dynamic> msg) {
  _message = msg['notification']['body'];
  print(_message);
},
onLaunch: (Map<String, dynamic> msg) {
  _message = msg['notification']['body'];
  print(_message);
},
```
## Result

All three messages are printed for Android. To be more clear, `onMessage` gets executed when the app is in the foreground when it gets a notification, `onResume` when the app is running in the background and you tap on the notification, `onLaunch` when the app isn't running and you tap on the notification. But on iOS, none of the `print` statements are executed. The notifications do arrive when the app is not in the foreground, but nothing happens when you click on them. No `print` outputs in XCode or VSCode. This leads to the idea that none of the three functions are getting called on iOS.

## How I solved it

On looking more closely, I found out that all three functions are actually getting called. Only, execution was apparently getting stopped at `_message = msg['notification']['body'];`, so the `print` never got called on iOS. Flutter doesn't log any complaints either.

The issue is, the path to the "Notification Text" is different on iOS and Android, so I ended up using this code:
```dart
onMessage: (Map<String, dynamic> msg) {
  if (Theme.of(context).platform == TargetPlatform.iOS)
      _message = msg['aps']['alert'];
  else
      _message = msg['notification']['body'];
  print(_message);
},
onResume: (Map<String, dynamic> msg) {
  if (Theme.of(context).platform == TargetPlatform.iOS)
      _message = msg['aps']['alert'];
  else
      _message = msg['notification']['body'];
  print(_message);
},
onLaunch: (Map<String, dynamic> msg) {
  if (Theme.of(context).platform == TargetPlatform.iOS)
      _message = msg['aps']['alert'];
  else
      _message = msg['notification']['body'];
  print(_message);
},
```
Now everything works on both iOS and Android.
## How this problem can be prevented
Going further, I request the Flutter/Firebase team to document how ```msg``` looks on iOS and Android. The ```firebase_messaging``` plugin's README is a good place. Also, why isn't Flutter throwing any error when it can't access ```msg['notification']['body']``` on iOS?
## Note
#23148 is somewhat different... that one is about keeping the structure of ```msg``` the same across `onLaunch, onResume` and `onMessage`. I actually find them to be the same! The issue I'm raising here is that it's not clear what the structures on Android and iOS are. It's desirable for the structures to be the same, but it's not as important as knowing what they are

**Comment by kbrmimbyl** [comment]
Digging further...

When I send a message from the Firebase console, with only the mandatory "Notification text", that text appears as `{aps: {alert: <notification text>}}` in the notification payload on iOS. This needs to be accessed as `msg['aps'][alert']`.

If I add the optional "Notification title" in the Firebase console, it becomes `{aps: {alert: {title: <notification title>, body: <notification text>}}`. In this case, the notification text needs to be accessed as `msg['aps']['alert']['body']` and the notification title as `msg['aps']['alert']['title']`.

On Android, it's consistently `msg['notification']['body']` irrespective of whether I add a title or not. If I do, the title is accessible as  `msg['notification']['title']`.

This means we need different code not only for iOS vs Android, but also title vs no title!

I did fire a cloud function to implement all this. In the cloud function, the above behavior is reproducible by using `{notification: {title: <Notification title>, 'body': <Notification text>}` as the payload argument of  `admin.messaging().sendToDevice()`.

**Comment by nayanbhana** [comment]
I have tried all of these steps and still can't get any of the hooks to fire.
ATM I currently will only see the push notification if the app is in the background.
When the app is in the background and a push notification comes through, I can click the notification and it will open the app but I see
`2019-02-21 08:13:35.652779+1100 Runner[312:8514] Warning: UNUserNotificationCenter delegate received call to -userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler: but the completion handler was never called.
`
 which leads me to believe my iOS setup is incorrect somewhere.

**Comment by nayanbhana** [comment]
What version of firebase messaging are you using @kbrmimbyl

**Comment by nayanbhana** [comment]
I have finally got the hooks to fire. I had to remove `flutter_local_notfications` from my project completely. I found this information here https://stackoverflow.com/questions/50970343/flutter-fcm-on-ios-not-calling-onmessage-and-onresume/51032413

**Comment by kbrmimbyl** [comment]
Sorry for the delay, @nayanbhana. Glad you got it working. I had trouble even installing any local notification plugin, so I didn't use any. My firebase_messaging plugin is ^3.0.1

**Comment by collinjackson** [comment]
Note that #23148 is now fixed, as a breaking change in 4.0.0. It seems like this issue is mostly about documentation of the functionality of the plugin, especially on iOS.

**Comment by kbrmimbyl** [comment]
@collinjackson "Now fixed"? I see 3.0.1 as the latest version on pub...

**Comment by danwulff** [comment]
I also have found that the notification model is different on iOS and Android when the optional data attribute is utilized. See #29027

**Comment by ghenry** [comment]
Anyone done a PR for the FCM plugin docs?

**Comment by cremor** [comment]
Has this been fixed by flutter/plugins#1728 or is there still something missing in the docs?

**Comment by divan** [comment]
> Has this been fixed by flutter/plugins#1728 or is there still something missing in the docs?
Thanks for the link - it points right to the part of README where this issue is explained. Technically speaking, this PR "fixes" the issue, but practically speaking – not really, as most users will probably copy-paste code from the `_firebaseMessaging.configure(..)` code block and assume that it's a good enough to start own data extraction from the message. It's unlikely to someone to assume that data structure will be different for iOS and Android, so 'Notification messages with additional data` part is skipped on the first read. Plus, people don't actually RTFM, as we know. :)
I, too spent many hours trying to understand what's wrong with my push notification handling logic, and found answer only in this thread, in the link to the PR, so maybe README can be improved.

**Comment by kroikie** [comment]
@kbrmimbyl

This issue has been moved to https://github.com/FirebaseExtended/flutterfire/issues/427. Any further collaboration will be done there.
